java-connectors/kafka-connect-azure-servicebus/src/main/java/io/lenses/streamreactor/connect/azure/servicebus/source/ServiceBusReceiverFacade.java

```diff
@@ -0,0 +1,157 @@
+    this.connectionString = connectionString;
+    this.receiverId = receiverId;
+    this.receiver = receiver;
+
+  }
+
+  private void instantiateReceiverFromKcql() {
+    Map<String, String> kcqlProperties = kcql.getProperties();
+    String subscriptionName = kcqlProperties.get(ServiceBusKcqlProperties.SUBSCRIPTION_NAME.getPropertyName());
+    String busType = kcqlProperties.get(ServiceBusKcqlProperties.SERVICE_BUS_TYPE.getPropertyName());
+
+    ServiceBusReceiverClientBuilder serviceBusReceiverClientBuilder =
+        new ServiceBusClientBuilder()
+            .connectionString(connectionString)
+            .receiver();
+
+    String outputTopic = kcql.getTarget();
+    String inputBus = kcql.getSource();
+
+    if (TOPIC_TYPE.equalsIgnoreCase(busType)) {
+      serviceBusReceiverClientBuilder.topicName(inputBus);
+      serviceBusReceiverClientBuilder.subscriptionName(subscriptionName);
+    } else if (QUEUE_TYPE.equalsIgnoreCase(busType)) {
+      serviceBusReceiverClientBuilder.queueName(inputBus);
+    }
+
+    receiver = serviceBusReceiverClientBuilder.buildAsyncClient();
+
+    subscription =
+        receiver.receiveMessages()
+            .doOnNext(onSuccessfulMessage(receiverId, recordsQueue, inputBus, outputTopic))
+            .doOnError(onError(receiverId))
+            .subscribe();
+  }
+
+  /**
+   * Disposes of subscription and closes the Receiver.
+   */
+  public void unsubscribeAndClose() {
+    try {
+      subscription.dispose();
+      receiver.close();
+    } catch (Exception e) {
+      log.error("{} - error while closing: {}", receiverId, e.getMessage());
+    }
+  }
+
+  private static Consumer<Throwable> onError(String receiverId) {
+    return error -> log.warn("{} - Error occurred while receiving message: ", receiverId, error);
+  }
+
+  void complete(ServiceBusReceivedMessage serviceBusMessage) {
+    receiver.complete(serviceBusMessage);
+  }
+
+  private static Consumer<ServiceBusReceivedMessage> onSuccessfulMessage(
+      String receiverId, BlockingQueue<ServiceBusMessageHolder> recordsQueue,
+      String inputBus, String outputTopic) {
+    return message -> {
+      long sequenceNumber = message.getSequenceNumber();
+      AzureServiceBusOffsetMarker offsetMarker =
+          new AzureServiceBusOffsetMarker(sequenceNumber);
+      AzureServiceBusPartitionKey partitionKey =
+          new AzureServiceBusPartitionKey(inputBus, message.getPartitionKey());
+
+      try {
+        SourceRecord sourceRecord = mapSingleServiceBusMessage(message, outputTopic, partitionKey, offsetMarker);
+        ServiceBusMessageHolder serviceBusMessageHolder =
+            new ServiceBusMessageHolder(message, sourceRecord, receiverId);
+        boolean offer = false;
+        while (!offer) {
+          offer = recordsQueue.offer(serviceBusMessageHolder, FIVE_SECONDS_TIMEOUT, TimeUnit.SECONDS);
+        }
+      } catch (InterruptedException e) {
+        log.info("{} has been interrupted on offering", receiverId);
```

Comment:
Should we consider additional handling for the InterruptedException beyond logging? For example, should we retry the operation, perform specific cleanup actions, or propagate the exception to maintain the application's interrupted state?

Reply:
Good question! For now:
1. If record won't reach recordsQueue it won't be sent to Kafka
2. Therefore it won't be committed in Service Bus (the logic to commit it won't work), which will mean we'll have to process it again anyway.

However, I still consider this a valid point. Log probably should be more visible, so I will definitely bump it's level to WARN (we can think if ERROR too) and maybe we can also consider timeout to be configurable?